shielder/contract/merkle.rs

```diff
@@ -0,0 +1,81 @@
+use ink::{
+    env::hash::{CryptoHash, Sha2x256},
+    storage::Mapping,
+};
+
+use crate::{
+    errors::ShielderError,
+    types::{Scalar, Set},
+};
+
+/// depth of the tree
+pub const DEPTH: usize = 10;
+
+#[ink::storage_item]
+#[derive(Debug)]
+pub struct MerkleTree {
+    /// mapping of tree indexes to values held in nodes
+    nodes: Mapping<u32, Scalar>,
+    /// set of historical roots (nodes[1]) of tree
+    roots_log: Set<Scalar>,
+    /// index of next available leaf
+    next_leaf_idx: u32,
+    /// number of leaves in the tree, should be equal to 2^DEPTH
+    size: u32,
+}
+
+pub fn compute_hash(first: Scalar, second: Scalar) -> Scalar {
+    let mut res = [0x0; 32];
+    Sha2x256::hash([first.bytes, second.bytes].concat().as_slice(), &mut res);
+    Scalar { bytes: res }
+}
+
+impl Default for MerkleTree {
+    fn default() -> Self {
+        Self {
+            nodes: Default::default(),
+            roots_log: Default::default(),
+            next_leaf_idx: 0,
+            size: (1 << DEPTH),
+        }
+    }
+}
+
+impl MerkleTree {
```

Comment:
a todo (or better, a Jira item) - write tests for this module, especially for the edge cases